core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/WindowsShareCopy.java

```diff
@@ -0,0 +1,276 @@
+    conf.setLong("io.file.buffer.size", config.bufferSize);
+    String[] list = dir.list();
+    int threadCount = (config.numThreads == null) ? 1 : config.numThreads;
+
+    // Copies the files in a multithreaded way
+    CountDownLatch executorTerminateLatch = new CountDownLatch(1);
+    ExecutorService executorService = createExecutor(threadCount, executorTerminateLatch);
+    CompletionService<String> completionService = new ExecutorCompletionService<>(executorService);
+
+    try {
+      for (final String file : list) {
+        completionService.submit(new Callable<String>() {
+          @Override
+          public String call() throws Exception {
+            String fileName = "";
+            try {
+              if (smbDirectory.endsWith("/")) {
+                fileName = copyFileToHDFS(hdfs, smbDirectory + file, hdfsDir, auth);
+              } else {
+                fileName = copyFileToHDFS(hdfs, smbDirectory + "/" + file, hdfsDir, auth);
+              }
+            } catch (IOException e) {
+              LOG.warn("Exception while copying the file {}", file, e);
+            }
+            return fileName;
+          }
+        });
+      }
+
+      for (int i = 0; i < list.length; i++) {
+        try {
+          Future<String> fileWritten = completionService.take();
+          String fileName = fileWritten.get();
+          LOG.info("{} is copied", fileName);
+        } catch (Throwable t) {
+          Throwable rootCause = Throwables.getRootCause(t);
+          if (rootCause instanceof ExecutionException) {
+            LOG.error("Exception occurred while copying the file.", rootCause);
+            throw (ExecutionException) rootCause;
+          }
+          if (rootCause instanceof InterruptedException) {
+            LOG.error("Copying aborted.", rootCause);
+            break;
+          }
+          Throwables.propagateIfPossible(rootCause, Exception.class);
+          throw Throwables.propagate(rootCause);
+        }
+      }
+    } finally {
+      executorService.shutdownNow();
+      executorTerminateLatch.await();
+    }
+  }
+
+  private String copyFileToHDFS(FileSystem hdfs, String smbSourceFile, Path dest, NtlmPasswordAuthentication auth)
+    throws IOException {
+    SmbFile smbFile = new SmbFile(smbSourceFile, auth);
+    String name = smbFile.getName();
+    Path nDest = new Path(dest, name);
+    LOG.debug("Thread {} is accessing source file {}, dest file {}", Thread.currentThread().getName(),
+              smbFile.getName(), nDest.getName());
+    // If file already exists, then we skip over.
+    if (hdfs.exists(nDest)) {
+      LOG.info("File {} already exists on HDFS, Skipping", nDest.getName());
+      return name;
+    }
+    LOG.info("Copying file {} to {}", smbSourceFile, nDest.toString());
+    try (InputStream in = smbFile.getInputStream();
+         BufferedOutputStream out = new BufferedOutputStream(hdfs.create(nDest))) {
+      ByteStreams.copy(in, out);
+    } catch (IOException e) {
```

Comment:
This doesn't seem like `INFO`.
Also, why isn't this exception thrown (instead of just caught and logged)? What's the behavior for other files that we want?